### ISSUE TITLE
typing warning with Chakra UI's `Input` on sample-app

### ISSUE THREAD

**Issue by rmobis** [airjp73/remix-validated-form#31]
Running the sample-app example on this repository, which uses Chakra UI, seems to be giving a `Type 'string' is not assignable to type 'never'. ts(2322)` warning on the `id` property at [`app/components/FormInput.jsx:30`](https://github.com/airjp73/remix-validated-form/blob/main/apps/sample-app/app/components/FormInput.tsx#L30). Including the snippet here for easier reading:

```tsx
// ...

  return (
    <>
      <FormControl isInvalid={!!error} isRequired={isRequired}>
        <FormLabel htmlFor={name}>{label}</FormLabel>
        <Input
          {...getInputProps({
            id: name,
            ...rest,
          })}
        />
        {error && <FormErrorMessage>{error}</FormErrorMessage>}
      </FormControl>
    </>
  );
```

Now, everything still seems to work fine and my experience with Chakra UI is none so I might be missing something obvious, but seeing as this is an official example I'm wondering if there's anything to be done about it...

**Comment by airjp73** [comment]
It looks like the type inference is not working correctly for that component. But it does work for a bare `input` which is super weird to me. 🤔 

If I change the `GetInputProps` type to this, then it works.
```ts
export type GetInputProps = <T extends Record<string, any>>(
  props: T
) => T;
```

As a worse-case scenario that's doable, but what the current type does that I want to keep is:
1. Removes `name` and `defaultValue` as props you can pass in (because you should use the ones from `useField`)
2. Ensures that `onChange` and `onBlur` are optional in case you're using a component that requires them.

Number 2 might be worth backtracking on. It seems pretty unlikely to have those be required, and if they are the input is probably not meant to be uncontrolled so you should probably be handling things yourself in that case.

Number 1 is a nice-to-have.

I'll play around with this a bit more later tonight.

**Comment by rmobis** [comment]
Not sure how useful this info might be, but I did some investigating so I thought I'd share what I found:

1. It's not only `id` that gives an error, any prop passed to `getInputProps` is expected to be `never` and throws the same error.
2. Apparently this happens because TypeScript matches matches the return type of `getInputProps` to the "return type" of Chakra Ui's `Input`, which is `ComponenWithAs<InputProps, "input">` also from Chakra.
3. When checked through the `node_modules` from, Chakra's own code has very similar `'<type>' is not assignable to type 'never'.` warnings on the `Input` component code (`node_modules\@chakra-ui\input\src\input.tsx:51` but when I downloaded their own repo, none is shown. Maybe it's something on their `tsconfig.json`?

For now, I'm working around this issue by just casting `name` to `never`:

```tsx
<Input
	{...getInputProps({
		id: name as never,
		...rest,
	})}
/>
```

**Comment by airjp73** [comment]
I found a solution and will publish it. Somehow switching from `Omit<T, 'defaultValue' | 'name'>` to `T & { defaultValue?: never, name?: never }` fixed it. One downside to this is that `defaultValue` and `name` show up as autocomplete options, but it still prevents you from setting them.

**Comment by airjp73** [comment]
The fix is live in 3.3.1. Feel free to reopen if that doesn't work. 😊

**Comment by airjp73** [comment]
@rmobis, I'm going to end up reverting that fix, I think. I'm working on adding to `getInputProps` and running into issues with the current typing. I've tried the old typing with other component libraries and Chakra seems to be the only one with this issue.

For an alternate fix, you can get the correct functionality without `as never` or `as any` is by supplying the generic directly.

```tsx
import { Input, InputProps } from '@chakra-ui/react';

<Input {...getInputProps<InputProps>({ id: name, ...rest })} />
```
